src/main.rs

```diff
@@ -0,0 +1,300 @@
+    /// started without immediate error but may not yet be ready to receive requests. The health status
+    /// endpoints are not available until after configuration has been read, so if this endpoint can
+    /// return a response at all, this value will be true.
+    pub live: bool,
+    /// The started field becomes true after the primary service is ready to receive requests. This occurs
+    /// after all plugins have been compiled but before any plugin is instantiated by an incoming request.
+    /// Once true, it will never become false.
+    pub started: bool,
+    /// The ready field indicates that the primary service is ready to receive requests. Theoretically,
+    /// this could become false if a service becomes unlikely to ever be able serve requests again in the
+    /// future, and should be set to false if the desired behavior is to trigger a restart. Currently the
+    /// ready field has identical behavior to the started field since no detection for things like
+    /// deadlock states exist yet.
+    pub ready: bool,
+}
+
+/// The default probe handler is intended to be at the apex of the health check resource. It simply performs
+/// a liveness health check by default.
+///
+/// See probe_handler.
+async fn default_probe_handler(
+    State(state): State<Arc<Mutex<HealthState>>>,
+) -> (StatusCode, Json<HealthResponse>) {
+    probe_handler(State(state), Path(String::from("live"))).await
+}
+
+/// The probe handler returns a JSON HealthResponse with a status code that depends on the probe type requested.
+///
+/// - live - Always returns an HTTP OK status if the endpoint is serving requests.
+/// - started - Returns an HTTP OK status if the primary service has started and is ready to receive requests
+///     and a Service Unavailable status otherwise.
+/// - ready - Returns an HTTP OK status if the primary service is available and ready to receive requests
+///     and a Service Unavailable status otherwise.
+async fn probe_handler(
+    State(state): State<Arc<Mutex<HealthState>>>,
+    Path(probe): Path<String>,
+) -> (StatusCode, Json<HealthResponse>) {
+    let state = state.lock().unwrap();
+    let status = match probe.as_str() {
+        "live" => StatusCode::OK,
+        "started" => {
+            if state.started {
+                StatusCode::OK
+            } else {
+                StatusCode::SERVICE_UNAVAILABLE
+            }
+        }
+        "ready" => {
+            if state.ready {
+                StatusCode::OK
+            } else {
+                StatusCode::SERVICE_UNAVAILABLE
+            }
+        }
+        // hint that the wrong probe value was sent
+        _ => StatusCode::NOT_FOUND,
+    };
+    (
+        status,
+        Json(HealthResponse {
+            live: true,
+            started: state.started,
+            ready: state.ready,
+        }),
+    )
+}
+
+/// An [`EnvFilter`] pattern to limit matched log events to error events.
+const ERROR_FILTER: &str = "error";
```

Comment:
Strum is new to me, but looks like a good solution for this.

Reply:
I think Strum matches the latter example significantly better than the former. In the case of the log level to filter mapping, it's not really a serialization mapping or a display format conversion but rather a mapping of two unlike things that happen to bear significant resemblance for most of their values. The `DEBUG_FILTER` makes the case for why they're different while the fall-through case further reinforces that it's not really enum-like since there are infinite possible values here.